### PR TITLE
Support shader preprocessor concatenation symbol

### PR DESCRIPTION
**Pull request by JohanAR**

### CODE REVIEW

File: servers/rendering/shader_preprocessor.cpp
Comment (JohanAR) [inline servers/rendering/shader_preprocessor.cpp:1077]: I removed the padding " " because they made unit testing much more complicated. Have not found any case where this caused problems.

File: servers/rendering/shader_preprocessor.cpp
Comment (JohanAR) [inline servers/rendering/shader_preprocessor.cpp:1274]: This is another thing I did because of unit testing. The preprocessor replaces #defines with empty lines. I don't know if this is intentional but I didn't want to mess with that. However I found that if the #define was indented, the preprocessor would leave a line containing only a space.

**Comment by clayjohn** [comment]
CC @bitsawer

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (JohanAR) [inline tests/servers/rendering/test_shader_preprocessor.h:181]: And in case anyone was wondering, this is my actual use case that motivated this PR. I wanted to make a sorting network more easily.

File: servers/rendering/shader_preprocessor.cpp
Comment (bitsawer) [inline servers/rendering/shader_preprocessor.cpp:1275]: I don't think this is a good place to try to format the output to make testing easier, it would be an endless whack-a-mole situation. You can remove this and adjust the testing to ignore lines that only have whitespace or whatever is needed, you can be relaxed like that in the testing comparison. The empty lines are kept to make side-by-side diffing easier for debugging. The spaces are there because get_token() will collapse any sequence of whitespace (spaces, tabs) into a single space and this is not the only place which can cause that, for example this should also cause similar issues:

``` glsl
#if DEBUG
    #define MEMBER_DEBUG_VARIABLE float debug_value;
#else
    #define MEMBER_DEBUG_VARIABLE
#endif

struct Data {
    float x;
    MEMBER_DEBUG_VARIABLE
};
```

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (bitsawer) [inline tests/servers/rendering/test_shader_preprocessor.h:36]: Unused include? Can be removed.

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (bitsawer) [inline tests/servers/rendering/test_shader_preprocessor.h:34]: Add `#include "tests/test_macros.h"` around here (clang-format might sort the order, so run it afterwards). It's technically not needed and things will compile because it is a header file that is pasted into test body later which has that include also, but missing that include confuses static analyzers and intellisense like tools in that file.

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (bitsawer) [inline tests/servers/rendering/test_shader_preprocessor.h:38]: Wrap rest of the file in a "namespace TestShaderPreprocessor {}", similar how (most?) other test files have their own namespaces.

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (bitsawer) [inline tests/servers/rendering/test_shader_preprocessor.h:154]: This TEST_CASE_MAY_FAIL() is bit ugly in the test output, printing a big red error and the failed comparison text in the console even though the full test suite still passes. It might be better to make it a normal TEST_CASE() and "fail succesfully" to keep the "noise" in the testing results output minimal. The comment there should be enough to clarify the test case.

**Review by bitsawer — CHANGES_REQUESTED** [review]
Overall, looks pretty good! Some small changes needed, but nothing fundamental. Gave it a few quick tests and nothing exploded. Don't worry about making those tests look ugly and full of whitespace with weird indentation and lax comparison rules. As discussed on RocketChat, I'm working on that shader preprocessor test suite so I'll add and format your tests into it and eventually try to make a PR once I have some extra time to finish it. I can give this another test later and a core team member will also have to give their review, but I don't see any major issues.

File: servers/rendering/shader_preprocessor.cpp
Comment (bitsawer) [reply servers/rendering/shader_preprocessor.cpp:1077]: Should be fine, like I mentioned in RocketChat adding those spaces might have been there "just in case".

File: servers/rendering/shader_preprocessor.cpp
Comment (bitsawer) [reply servers/rendering/shader_preprocessor.cpp:1274]: Commented this in my review.

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (JohanAR) [reply tests/servers/rendering/test_shader_preprocessor.h:154]: Before I discovered the TEST_CASE_MAY_FAIL macro I had it marked both as may_fail and skip, but I'll do as you suggest.

File: servers/rendering/shader_preprocessor.cpp
Comment (JohanAR) [reply servers/rendering/shader_preprocessor.cpp]: Removed it and made the test comparator more lenient.

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (JohanAR) [reply tests/servers/rendering/test_shader_preprocessor.h]: Done

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (JohanAR) [reply tests/servers/rendering/test_shader_preprocessor.h]: Done

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (JohanAR) [reply tests/servers/rendering/test_shader_preprocessor.h:48]: Done

**Comment by Calinou** [comment]
Remember to document this change in https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/shader_preprocessor.html :slightly_smiling_face:

File: servers/rendering/shader_preprocessor.cpp
Comment (JohanAR) [reply servers/rendering/shader_preprocessor.cpp:1077]: I found one case where it changes the behaviour:

```
#define MOD +
float a MOD= 1.0;
```

Previously this would've been expanded to invalid code, while now it works. I did some tests with a C++ compiler and it doesn't work there.

Should I revert this line and update the tests to require the surrounding spaces? I think it would be difficult to relax the test to allow them without adding a full blown glsl parser/tokenizer.

**Comment by JohanAR** [comment]
Docs PR https://github.com/godotengine/godot-docs/pull/6979

**Review by bitsawer — APPROVED** [review]

**Comment by bitsawer** [comment]
Looks good to me after the changes. Tested with my own projects and tests, seems to work as expected. I'll add those tests to my preprocessor test suite and try to eventually make a public PR for it once I have some extra time.

@Chaosus might still want to take a look and give the final approval.

File: servers/rendering/shader_preprocessor.cpp
Comment (Chaosus) [inline servers/rendering/shader_preprocessor.cpp:1133]:
```suggestion
		int index_end = index_start + 2; // First character after ##.
		// The macro was checked during creation so this should never happen.
```

File: servers/rendering/shader_preprocessor.cpp
Comment (Chaosus) [inline servers/rendering/shader_preprocessor.cpp:1147]:
```suggestion
		// Skip whitespace after ##.
```

File: servers/rendering/shader_preprocessor.cpp
Comment (Chaosus) [inline servers/rendering/shader_preprocessor.cpp:1152]:
```suggestion
		// Skip whitespace before ##.
```

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (Chaosus) [inline tests/servers/rendering/test_shader_preprocessor.h:59]:
```suggestion
// Remove unnecessary spaces from a line.
```

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (Chaosus) [inline tests/servers/rendering/test_shader_preprocessor.h:62]:
```suggestion
	// Result is guaranteed to not be longer than the input.
```

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (Chaosus) [inline tests/servers/rendering/test_shader_preprocessor.h:69]:
```suggestion
		// These test cases only use ASCII.
```

File: tests/servers/rendering/test_shader_preprocessor.h
Comment (Chaosus) [inline tests/servers/rendering/test_shader_preprocessor.h:76]:
```suggestion
				// E.g. "float x" or "- -".
```

**Review by Chaosus — APPROVED** [review]
Tested and it works fine for me - just add the missing dots to the commentaries.

**Comment by JohanAR** [comment]
> Tested and it works fine for me - just add the missing dots to the commentaries.

Thanks! I'm not that used to working with github's web UI, but I think I got all the suggestions. Should I try to squash the branch or will that be done when merging?

**Comment by Chaosus** [comment]
> Should I try to squash the branch or will that be done when merging?

Yes, you should.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
